pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -471,13 +468,13 @@ func hasFinalizer(accessor metav1.Object, matchingFinalizer string) bool {
 // this function takes newAccessor directly because the caller already
 // instantiates an accessor for the newObj.
 func startsWaitingForDependentsDeleted(oldObj interface{}, newAccessor metav1.Object) bool {
-	return deletionStarts(oldObj, newAccessor) && hasDeleteDependentsFinalizer(newAccessor)
+	return deletionStarts(oldObj, newAccessor, func(a metav1.Object) bool { return hasDeleteDependentsFinalizer(a) }) && hasDeleteDependentsFinalizer(newAccessor)
```

Comment:
This is pretty confusing to follow... would renaming to deletionStartsWithFinalizer(oldObj, newAccessor, finalizer) and moving all the logic into that method be easier to follow?

```
// if the new object isn't being deleted, or doesn't have the finalizer we're interested in, return false
if !beingDeleted(newAccessor) || !hasFinalizer(newAccessor, finalizer) {
  return false
}

// if the old object is nil, or wasn't being deleted, or didn't have the finalizer, return true
if oldObj == nil {
  return true
}
oldAccessor, err := meta.Accessor(oldObj)
if err != nil {
  utilruntime.HandleError(fmt.Errorf("cannot access oldObj: %v", err))
  return false
}
return !beingDeleted(oldAccessor) || !hasFinalizer(oldAccessor, finalizer)
```

Reply:
it looks much better, thanks Jordan :)